src/core/net/ip6.cpp

```diff
@@ -1121,8 +1121,10 @@ Error Ip6::HandleDatagram(OwnedPtr<Message> aMessagePtr, const void *aLinkMessag
         }
 #endif
 
-        // Always forward multicast packets to host network stack
-        forwardHost = true;
+        // Always forward multicast packets to host network stack unless packet is sent from
+        // link-local or mesh-local address to multicast group with scope larger than realm-local
+        forwardHost = !(header.GetDestination().IsMulticastLargerThanRealmLocal() &&
+                        (header.GetSource().IsLinkLocal() || Get<Mle::Mle>().IsMeshLocalAddress(header.GetSource())));
```

Comment:
Consider there's a Thread network without an infra link. I think it's valid for a Thread node to create a socket on the host listening to a multicast group whose scope is larger than realm-local. In such a case the node who sends packets to this multicast group will use a mesh-local address. This filter will drop such packets. Is this acceptable?

Reply:
@bukepo @abtink  thoughts? Does this drop the use case you wanted to support?

Reply:
From https://www.ietf.org/rfc/rfc3484.txt, it looks like allowed to use smaller scope source address for multicast traffic. So the packet should be delivered to the host allowing host stack's endpoints to process it. However, it should not be forwarded out of the scope of its source address, so I think should defer the restrict to the host stack.

Reply:
SGTM as a temporary solution. However, mangling the packet looks like a unexpected behavior here. I hope we can implement the filtering on Android host in future. Given this is only required for the current Android limitation, should we add a compile time flag or guide it by existing android flags?

Reply:
Sounds good! Using the __ANDROID__ flag to limit this change only for Android.